### PR TITLE
feat: add siphash24 implementation

### LINKED ISSUE #2: Implement siphash24
U can watch how it implemented in [rust](https://github.com/rust-bitcoin/rust-bitcoin/blob/0123f2894c66419f1ad561e9e7d98af66c4db9ab/hashes/src/siphash24.rs#L339).

We expecting from this PR:
 * hashing algorithm in Zig style (u can watch std.crypto.hash.sha2 family or other)
 * tests

### PR DESCRIPTION
**Pull request by TropicalDog17**
## Description

Closes #2 

\cc @StringNick

### CODE REVIEW

File: src/hashes/siphash24.zig [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+            var buf = [_]u8{0} ** 8;
+            @memcpy(buf[0..b.len], b);
+            buf[7] = self.msg_len;
+            self.round(buf);
+
+            if (T == u128) {
+                self.v2 ^= 0xee;
+            } else {
+                self.v2 ^= 0xff;
+            }
```
Comment (tcoratger) [inline src/hashes/siphash24.zig:69]:
```suggestion
           self.v2 ^=  if (T == u128) 0xee else 0xff;
```

File: src/hashes/siphash24.zig [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+        pub const d_rounds = 4;
+        const State = SipHashStateless(T, c_rounds, d_rounds);
+
+        state: State,
+        buf: [8]u8,
+        buf_len: usize,
+
+        /// Initialize a state for a SipHash function
+        pub fn init(key: *const [key_length]u8) Self {
+            return Self{
```
Comment (tcoratger) [inline src/hashes/siphash24.zig:148]:
```suggestion
            return .{
```

File: src/hashes/siphash24.zig [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+            const aligned_len = msg.len - (msg.len % 8);
+            var c = Self.init(key);
+            @call(.always_inline, update, .{ &c, msg[0..aligned_len] });
+            return @call(.always_inline, final, .{ &c, msg[aligned_len..] });
+        }
+    };
+}
+
+fn SipHash24(comptime T: type) type {
+    assert(T == u64 or T == u128);
```
Comment (StringNick) [inline src/hashes/siphash24.zig:131]: i think here better to use @compileError

Comment (StringNick) [inline src/hashes/siphash24.zig:131]: i am not sure, maybe better compileError? wdyt @tcoratger @TropicalDog17

File: src/hashes/siphash24.zig [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+
+    return struct {
+        const Self = @This();
+        pub const key_length = 16;
+        pub const mac_length = @sizeOf(T);
+        pub const block_length = 8;
+        pub const c_rounds = 2;
+        pub const d_rounds = 4;
+        const State = SipHashStateless(T, c_rounds, d_rounds);
```
Comment (StringNick) [inline src/hashes/siphash24.zig:140]: if we support u64 / u128 and SipHashStateless same (as in std), maybe we can use  
```zig
const State = if (T == u64) SipHash64(c_rounds, d_rounds) else SipHash128(c_round, d_rounds);
```
and we dont need copy full SipHashStateless

File: src/hashes/siphash24.zig (hunk 3 above)
Comment (tcoratger) [reply src/hashes/siphash24.zig]: yeah probably better here

**Pull request merged**